src/odemis/driver/tmcm.py

```diff
@@ -2502,3 +2553,871 @@ def _parseMessage(self, msg):
+        # DISABLE" before the loop condition is evaluated, it will attempt a transition from "SWITCHED ON DISABLE" to
+        # "QUICK STOP ACTIVE". This transition is not allowed, so it will raise an error.
+        # To avoid this, we change the settings to disable the automatic change to "SWITCH ON DISABLED" state
+        # and do it manually in the StopAxis() function.
+        self.SetAxisParam(axis, QUICK_STOP_OPTION, 6)  # stay in quick stop active state
+
+    # Low level functions
+    def GetVersion(self, axis=0):
+        """
+        return (int, int, int):
+             Controller ID
+             Firmware major version number
+             Firmware minor version number
+        """
+        cont = self.GetAxisParam(axis, DEVICE_NAME)
+        vmaj = self.GetAxisParam(axis, HW_VERSION)  # e.g. '1.0'
+        vmin = self.GetAxisParam(axis, IDENTITY, 3)
+        return cont, int(float(vmaj)), int(vmin)
+
+    def GetAxisParam(self, axis, param, idx=None):
+        """
+        Read the axis/parameter setting from the RAM
+        axis (0<=int<=5): axis number
+        param (0<=int<=255): parameter number
+        return (0<=int): the value stored for the given axis/parameter
+        idx in case of record object
+        """
+        if idx:
+            val = self._node.sdo[param + 0x800 * axis][idx].raw
+        else:
+            val = self._node.sdo[param + 0x800 * axis].raw
+        return val
+
+    def SetAxisParam(self, axis, param, val):
+        """
+        Write the axis/parameter setting from the RAM
+        axis (0<=int<=3): axis number
+        param (0<=int<=255): parameter number
+        val (int): the value to store
+        """
+        # Axis 0 start with 0, axis 1 with 0x800, etc.
+        self._node.sdo[param + 0x800 * axis].raw = val
+
+    def MoveAbsPos(self, axis, pos):
+        """
+        Requests a move to an absolute position. This is non-blocking.
+        axis (0<=int<=5): axis number
+        pos (-2**31 <= int 2*31-1): position
+        """
+        self._node.controlword = 0x0f  # switch on
+        self.SetAxisParam(axis, TARGET_POSITION, pos)
+        # Bit 6 of controlword: absolute (0) or relative (1)
+        # Bits 0-3 to 1: switch on and enable (see p 71 of canopen firmware manual)
+        # Bit 4: start positioning
+        self._node.controlword = 0x1f
+
+    def MoveRelPos(self, axis, offset):
+        """
+        Requests a move to a relative position. This is non-blocking.
+        axis (0<=int<=5): axis number
+        offset (-2**31 <= int 2*31-1): relative position
+        """
+        self.SetAxisParam(axis, TARGET_POSITION, offset)
+        # Bit 6 of controlword: absolute (0) or relative (1)
+        # Bits 0-3 to 1: switch on and enable (see p 71 of canopen firmware manual)
+        # Bit 4: start positioning
+        self._node.controlword = 0xf  # switch on (transition to state 3)
+        self._node.controlword = 0b1011111
```

Comment:
Yes, the way it works is that there are different states that the hardware can be in. We have to write the controlword twice, first, to get into "switched on" state, then, to get into "operation enabled" state. I'll add a comment to make it more clear.

Reply:
I see. IMHO, that's one case where Python properties should not be used. If it was a function like ".setControlState()" it would look a lot less suspicious to call it twice with different values. Anyway, let's leave this API as it is ;-)